packages/swing-store/test/test-state.js

```diff
@@ -70,27 +69,44 @@ test('in-memory kvStore read/write', t => {
   testKVStore(t, initSwingStore(null));
 });
 
-test('persistent kvStore read/write/re-open', t => {
-  const dbDir = 'testdb';
-  t.teardown(() => rimraf(dbDir));
-  rimraf(dbDir);
+test('persistent kvStore read/write/re-open', async t => {
+  const [dbDir, cleanup] = await tmpDir('testdb');
+  t.teardown(cleanup);
   t.is(isSwingStore(dbDir), false);
   const store = initSwingStore(dbDir);
   const { commit, close } = store;
   testKVStore(t, store);
-  commit();
+  await commit();
   const before = getAllState(store);
-  close();
+  await close();
   t.is(isSwingStore(dbDir), true);
 
   const store2 = openSwingStore(dbDir);
   const { close: close2 } = store2;
   t.deepEqual(getAllState(store2), before, 'check state after reread');
   t.is(isSwingStore(dbDir), true);
-  close2();
+  await close2();
+});
+
+test('persistent kvStore maxKeySize write', async t => {
+  // Vat collections assume they have 220 characters for their key space.
+  // This is based on previous math where LMDB's key size was 511 bytes max
+  // and the native UTF-16 encoding of JS strings, minus some overhead
+  // for vat store prefixes.
+  // However some unicode codepoints may end up serialized as 3 bytes with UTF-8
+  // This tests that no matter what, we can write 254 unicode characters in the
+  // 0x0800 - 0xFFFF range (single UTF-16 codepoint, but 3 byte UTF-8).
+
+  const [dbDir, cleanup] = await tmpDir('testdb');
+  t.teardown(cleanup);
+  t.is(isSwingStore(dbDir), false);
+  const store = initSwingStore(dbDir);
+  store.kvStore.set('€'.repeat(254), 'Money!');
```

Comment:
Let's assert both sides of the size limit (i.e. assert that 255 3-byte-UTF-8 chars fails), since the success/failure of writes needs to be part of consensus. This isn't an exact constraint (we'd need a mix of encoded sizes, plus a healthy variety of surrogate pairs, to exercise this thoroughly), but I think it's good to be in the habit of knowing what the precise rules are.

I think that means this DB change is consensus-breaking (there are non-trivial unicode strings which would pass the earlier encoding scheme's limit but not the new one, or vice versa), but I'm ok with making that now.. better than later.

Reply:
> 255 3-byte-UTF-8 chars fails

It won't fail. The upper limit is actually 1971 bytes I believe.

> I think that means this DB change is consensus-breaking

I don't believe so. Nothing is writing large keys since the only user controlled key size is the collections keys which is independently checked.

Reply:
Ah, ok, so the previous limits were (pardon my js/python pseudocode) `str.length < X && len(str.encode('utf-16')) < Y`, and the new limits are `str.length < X && len(str.encode('utf-8') < Z`, but for all possible strings, the `.length < X` limit (imposed by liveslots) is tighter than the encoded limit (imposed lower down). So landing this can't make any failing userspace behavior start to succeed, or succeeding userspace behavior start to fail.

As long as that criteria is met, I guess we don't need more stringent testing at this layer.

Reply:
Yes, this test is specifically about checking this assumption. That the worst case `str = threeByteUtfChar.repeat(X)` then `len(str.encode('utf-8')) < Z`. Aka `3 * X < Z`